### PR TITLE
New system for keying listeners

### PR DESCRIPTION
**Pull request by elihart**
There has been a lot of issues and confusion, both within airbnb and externally, because of the `DoNotHash` option and how to deal with callbacks that capture state that may change.

This PR takes the approach suggested in this issue: https://github.com/airbnb/epoxy/issues/414

There is a `KeyedListener` class that wraps any kind of callback, and implements equals/hashcode with the value of a given key.

To make this easy to use I've written several helper functions in kotlin - java would make dealing with the generics a big pain, and kotlin also lets us do some nice things with reified type checking and extension functions. This is the first time I'm adding kotlin to the core module, but I think it's about time :P 

Note, I haven't written much code comments yet to avoid wasting time if the design changes, so consider it a WIP.

Usage for creating the listeners looks like this:
```kotlin
CarouselModel_().apply {

        // Click listener keyed on model
        onClickListener(keyedClickListener {

        })

        // Long click listener keyed on model
        keyedLongClickListener {
            return@keyedLongClickListener false
        }

        // Custom listener type, keyed on model
        keyedListener {
            4
        }

        // Custom listener type, keyed on user defined key
        keyedListener("custom key") {

        }

        // Custom interface callback, keyed on custom key
        keyedListener("some key", object : MyCustomInterface {
            override fun doSomething(view: View) {

                // Get details on the model and view
                val (model, parentView, clickedView, position) = clickContext(view) ?: return
            }
        })
    }
```

**High level points**
- Helpers like `keyedClickListener` create a click listener that is keyed on the model value, excluding listeners. So if the model changes at all the click listener is considered different and will be updated
- There are other helpers for using custom lambdas or interfaces as keyed callbacks.
- A custom key can be used if you don't want to key on the model
- Keyed listeners can access the `ClickContext` to get the most up to date model object, view, and adapter position. This can be accessed via the clicked view, or for default click listeners it is the lambda receiver for easy access
- I've updated the `@CallbackProp` annotation to only apply DoNotHash behavior if the type does not have equals/hashcode. This means you can use a KeyedListener with CallbackProp and it will be updated as expected

**Issues**
- This adds another way to handle click listeners, on top of the previous `OnModelClickListener` support. This is a bit confusing.
- If `@CallbackProp` is used you still need to explicitly define the listener type as KeyedClickListener for the hashing to work (same for databinding)
- The extension functions currently operate on the EpoxyModel as a receiver, so they won't work with the model builder extension functions. I would need to generate other methods for those to work

In general there is a lot of knowledge and overhead required to choose the right option and make sure your code has the right behavior. I think the only way to solve this is with a fairly drastic breaking change of how the generated models use listeners - that is maybe possible in the future, but this seems like a good step for now.

@ngsilverman @chrisbanes

### CODE REVIEW

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/KeyedListeners.kt [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        model?.inKeyedListenerEquals = true
+
+        val result = key.hashCode()
+
+        model?.inKeyedListenerEquals = false
+        return result
+    }
+}
+
+data class ClickContext<V : View, P : Any, T : EpoxyModel<P>>(
```
Comment (ngsilverman) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/KeyedListeners.kt]: The name "ClickContext" confused me a bit at first, until I understood that it's not directly related to clicks. It's just context about a view, and could be used at any time, including when non-click events happen, right?

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/KeyedListeners.kt [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+inline fun <reified V : Any, reified T : EpoxyModel<V>> T.keyedClickListener(
+    crossinline listener: ClickContext<View, V, T>.() -> Unit
+) = KeyedClickListener(this, View.OnClickListener { v ->
+    ClickContext.from<View, V, T>(v)?.apply(listener)
+})
+
+class KeyedClickListener(
+    key: Any,
+    listener: View.OnClickListener
+) : KeyedListener<View.OnClickListener>(key, listener), View.OnClickListener {
```
Comment (ngsilverman) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/KeyedListeners.kt]: Delegation could help here, and for other keyed listener types:
```kotlin
class KeyedClickListener(
    key: Any,
    listener: OnClickListener
) : KeyedListener<OnClickListener>(key, listener), OnClickListener by listener
```

**Comment by ngsilverman** [comment]
@elihart an additional thought: it seems like we could solve this nicely if we knew which values are captured by the listener closures. Since I don't think we can get them automatically, maybe we could make those the keys:
```kotlin
CarouselModel_().apply {
    onClickListener(keyedClickListener(listOf(stateA, stateB) {
        // Does something with stateA and stateB.
    }))
}
```

**Comment by elihart** [comment]
@ngsilverman thanks for looking, I guess I should give more background

> When using a listener keyed on the model itself, wouldn't it only be rebound if another model property changes? In which case it would have been rebound anyway.

The problem is that epoxy only rebinds properties that change, so if another property on the model changes it gets rebound, but not the click listener - in the current system the only way to have the listener rebound is if the hashcode changes, which doesn't happen because of DoNotHash - or otherwise it would always be rebound for anonymous click listeners.

Issues have been brought up like 
https://github.com/airbnb/epoxy/issues/502
https://github.com/airbnb/epoxy/issues/499

and someone at airbnb just messaged me wondering why their code wasn't working and they were doing this
```
onClickListener { v ->
                MvRxActivity.startActivity<UserProfileFragment>(v.context, UserProfileArguments(state.userId))
            }
```

the userid was changing and the click listener had captured the old one.

Another example is if you have two different listener objects and you want to swap between the two, that's hard to do right now.

It just occurred to me that another approach could be to have epoxy's internal binding rebind @CallbackProps if any non callbackprop changed - same idea as I have now in the keyed listener, but it would be more built in. However, there are still subtle cases, like I posted above, where the model state doesn't change but we still need the click listener to change, which is really hard to automatically detect and fix.

I'm still searching for a perfect solution that catches all cases and isn't confusing - not sure if that is a white whale though

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/KeyedListeners.kt [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    override fun onClick(v: View) = listener.onClick(v)
+}
+
+open class KeyedListener<Listener>(private val key: Any, val listener: Listener) {
+
+    // Only include the key, and not the listener, in equals/hashcode
+    override fun equals(other: Any?): Boolean {
+        val model = key as? EpoxyModel<*>
+        if (model?.inKeyedListenerEquals == true) return true
```
[the remark this answers is not in the document]
Comment (ngsilverman) [reply epoxy-adapter/src/main/java/com/airbnb/epoxy/KeyedListeners.kt]: @elihart could this cause issues if `equals` runs concurrently?

**Pull request closed without merging**